src/omt/bitvector_optimizer.cpp

```diff
@@ -0,0 +1,248 @@
+/*********************                                                        */
+/*! \file bitvector_optimizer.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Yancheng Ou
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ ** in the top-level source directory and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Optimizer for BitVector type
+ **/
+
+#include "bitvector_optimizer.h"
+
+#include "options/smt_options.h"
+#include "smt/smt_engine.h"
+#include "theory/quantifiers/quantifiers_attributes.h"
+#include "theory/smt_engine_subsolver.h"
+
+using namespace CVC4::theory;
+
+namespace CVC4::smt {
+
+OMTOptimizerBitVector::OMTOptimizerBitVector(bool isSigned)
+    : d_isSigned(isSigned)
+{
+}
+
+BitVector OMTOptimizerBitVector::computeAverage(const BitVector& a,
+                                                const BitVector& b,
+                                                bool isSigned)
+{
+  // computes (a + b) / 2 without overflow
+  // rounding towards -infinity: -1.5 --> -2,  1.5 --> 1
+  // average = (a / 2) + (b / 2) + (((a % 2) + (b % 2)) / 2)
+  uint32_t aMod2 = (uint32_t)(a.isBitSet(0));
+  uint32_t bMod2 = (uint32_t)(b.isBitSet(0));
+  BitVector aMod2PlusbMod2(a.getSize(), uint32_t((aMod2 + bMod2) / 2));
+  BitVector bv1(a.getSize(), (uint32_t)1);
+  if (isSigned)
+  {
+    return (a.arithRightShift(bv1) + b.arithRightShift(bv1)
+            + aMod2PlusbMod2.arithRightShift(bv1));
+  }
+  else
+  {
+    return (a.logicalRightShift(bv1) + b.logicalRightShift(bv1)
+            + aMod2PlusbMod2.logicalRightShift(bv1));
+  }
+}
+
+std::unique_ptr<SmtEngine> OMTOptimizerBitVector::createOptChecker(
```

Comment:
Similar to the integer case, it might be better to use the `createOptCheckerWithTimeout` method directly and delete this method.

Reply:
Resolved